python/pyarrow/tests/test_array.py

```diff
@@ -710,9 +710,9 @@ def test_struct_from_arrays():
     assert arr.to_pylist() == []
 
     # Inconsistent fields
-    fa2 = pa.field("a", pa.int32())
-    with pytest.raises(ValueError, match="int64 vs int32"):
-        pa.StructArray.from_arrays([a, b, c], fields=[fa2, fb, fc])
```

Comment:
I have to investigate further but think there's something wrong here. This use case should still raise an error as the underlying `a` child array is int64. I am not entirely sure if we are making a copy or what is happening, as I am not familiar with this area of the codebase, but this should raise an error AFAIU.

Reply:
Right, I am coercing inputs to match the provided types specified by the user. I don't think it's obvious whether this is right or wrong; there is no specification here.

We could require types to be identical (no coercion) if the input values are all Pyarrow Arrays. The implementation will just be a little more complex. Let me know what you prefer.